accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,498 @@
+
+The rationale for making apphost EXEs opt-in is that they are a non-portable asset, creating a strange asymmetry in an otherwise portable app. Perhaps, we should make it easy for developers to generate an apphost only for debug builds to aid the development experience. That said, if developers want a better development experience, they could just as easily build RID-specific apps (the proposed default) for development and then build a portable app with no EXE for production where an EXE is presumably not needed.
+
+Note: Portable apps are not a meaningful option for client apps since almost all client apps require an executable launcher, and there is only ever one generated. Portable apps are only meaningful for commandline apps.
+
+Note: In theory, we could consider an option to create multiple apphosts for different RIDs, either of different names or in different directories. Again, this capability would need to be opt-in.
+
+### Rationalize RID folders for RID-specific and portable apps
+
+There have been some issues filed (see appendix) that suggest that managed libraries are not resilient to RID-specific libraries changing location as a result of changing build type (portable to RID-specific). This would also be true for single file deployment builds.
+
+We could enable (as an option) creating `runtimes` folders for RID-specific apps to make the folder layout the same across deployment types. Doing that would put single file at a significant disadvantage. We'd rather encourage the ecosystem to support both folder layouts. If we had to choose one of these deployments types to disadvantage, it would be portable (with the `runtimes` folders), not RID-specific, which is the likely the opposite of what developers are tuning for today.
+
+We need to invest more in docs and tools to the end of clear and actionable guidance.
+
+### Use SDK RID as default RID
+
+RID-specific apps should default to the SDK RID as an implicit RID, and no longer require a RID to be specified. The SDK already does this for the apphost for portable apps. This proposal is a straightforward extension of that behavior.
+
+This means that the following commands would produce assets in terms of the implicit RID, such as with the following commands.
+
+- `dotnet publish --self-contained`
+- `dotnet publish /p:PublishSingleFile=true`
+
+Note: As already suggested, this change would make Dockerfiles MUCH easier to write. Needing to specify RIDs is probably the worst part of using .NET with Docker.
+
+Specifying a RID should have no effect on the build type/deployment model. It should simply refine the existing model. This change largely follows from using the SDK RID as the implicit default for all types of RID-specific apps. When a RID is specified, it is substituted for the implicit RID.
+
+This means that the following commands would produce a RID-specific framework-dependent app:
+
+- `dotnet build -r linux-x64`
+- `dotnet publish -r linux-x64 /p:PublishReadyToRun=true`
```

Comment:
Please also consider how NuGet lock files should work. Today, the following commands result in an error: https://github.com/NuGet/Home/issues/9195

```text
dotnet new console
dotnet restore --use-lock-file
dotnet restore --locked-mode -r linux-x64
```

While that looks like a contrived example (why would someone restore locked mode for a specific RID?), consider that customers want to use locked mode on CI build pipelines (via Directory.Build.props or environment variable), and that `dotnet publish -r linux-x64` does an implicit restore. Bang, restore failed, so publishing failed, and CI failed.

Customers want to use lock files and locked mode so they can use floating versions, but still get consistent restores across all machines, and also some package substitution prevention, as lock files ensure the package's content hash is the same when packages are extracted.

There needs to be a way for the SDK to ensure that the set of packages NuGet is asked to restore (plus the set of RIDs that NuGet is informed about) is consistent, regardless of what RID is being built/published by the comment line arguments.

This spec seems to be a higher level than these details, but lock files need to be designed to be workable from the top. NuGet can't have a workable lock file experience when the SDK reports a different set of packages on different actions.

Reply:
This is a good issue. I hadn't thought about that. Fortunately, this change doesn't come until .NET 7 so we'll have time properly design this experience. I agree with you.